pkg/mimirpb/symbols_test.go

```diff
@@ -0,0 +1,238 @@
+		s := NewFastSymbolsTable(baseSymbolsMapCapacity)
+		s.ConfigureCommonSymbols(8, commonSymbols)
+
+		require.Equal(t, []string{""}, s.Symbols(), "required empty reference does not exist")
+		require.Equal(t, uint32(0), s.Symbolize(""))
+		require.Equal(t, []string{""}, s.Symbols())
+
+		require.Equal(t, uint32(9), s.Symbolize("abc"))
+		require.Equal(t, []string{"", "abc"}, s.Symbols())
+
+		require.Equal(t, uint32(1), s.Symbolize("__name__"))
+		require.Equal(t, []string{"", "abc"}, s.Symbols())
+
+		require.Equal(t, uint32(10), s.Symbolize("def"))
+		require.Equal(t, []string{"", "abc", "def"}, s.Symbols())
+
+		require.Equal(t, uint32(2), s.Symbolize("__aggregation__"))
+		require.Equal(t, []string{"", "abc", "def"}, s.Symbols())
+
+		s.Reset()
+		require.Nil(t, s.commonSymbols)
+	})
+}
+
+func desymbolizeLabelsDirect(labelRefs []uint32, symbols []string) ([]LabelAdapter, string) {
+	name := ""
+	las := make([]LabelAdapter, len(labelRefs)/2)
+	for i := 0; i < len(labelRefs); i += 2 {
+		las[i/2].Name = symbols[labelRefs[i]]
+		las[i/2].Value = symbols[labelRefs[i+1]]
+		if las[i/2].Name == labels.MetricName {
+			name = las[i/2].Value
+		}
+	}
+	return las, name
+}
+
+func BenchmarkSymbolizer(b *testing.B) {
+	b.Run("2k labels unique values, common symbols", func(b *testing.B) {
+		lbls := make([]string, 2*2000)
+		for i := range 2000 {
+			lbls[i*2] = "__name__"
+			lbls[(i*2)+1] = fmt.Sprintf("series_%d", i)
+		}
+
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			st := symbolsTableFromPool()
+			st.ConfigureCommonSymbols(64, benchmarkCommonSymbols)
+
+			for _, l := range lbls {
+				_ = st.Symbolize(l)
+			}
+
+			symbols := symbolsSliceFromPool()
+			symbols = st.SymbolsPrealloc(symbols)
+			if len(symbols) != 2001 {
+				b.Fatalf("unexpected number of symbols: %d", len(symbols))
+			}
+			reuseSymbolsSlice(symbols)
+			reuseSymbolsTable(st)
+		}
+	})
+
+	b.Run("7 series with some shared label keys, common symbols", func(b *testing.B) {
+		numSeries := 7
+		generatedLabels := 15
+		series := [][]string{}
+		gen := rand.New(rand.NewSource(789456123))
+		for i := range numSeries {
+			s := []string{"__name__", fmt.Sprintf("series_%d", i), "namespace", "my-namespace", "job", fmt.Sprintf("namespace/job-%d", i)}
+			for i := range generatedLabels {
+				s = append(s, fmt.Sprintf("generated_label_%d", i), fmt.Sprintf("%d", gen.Uint64()))
+			}
+			series = append(series, s)
+		}
+
+		b.ResetTimer()
+
+		for n := 0; n < b.N; n++ {
+			st := symbolsTableFromPool()
+			st.ConfigureCommonSymbols(64, benchmarkCommonSymbols)
+
+			for _, s := range series {
+				for _, l := range s {
+					_ = st.Symbolize(l)
+				}
+			}
+
+			symbols := symbolsSliceFromPool()
+			symbols = st.SymbolsPrealloc(symbols)
+			if len(symbols) != 136 {
+				b.Fatalf("unexpected number of symbols: %d", len(symbols))
+			}
+			reuseSymbolsSlice(symbols)
+			reuseSymbolsTable(st)
+		}
+	})
+}
+
+var (
+	benchmarkCommonSymbols = []string{
```

Comment:
copied due to package dependency cycles. it's just local to a benchmark, we don't really care much if it drifts